### PR TITLE
feat: upgradeable contract

### LINKED ISSUE #11: feat: Upgradeability in Cairo
Starknet has a native syscall called `replace_class` that allows you to replace the class hash of the current contract, while keeping the same address. As such, the proxy pattern is no longer required on Starknet. 

In this example, we demonstrate a basic contract that can be upgraded after deployment with a new class hash.

### PR DESCRIPTION
**Pull request by enitrat**
Closes #11
@Julio4 take a look and let me know if you find it easy to understand :)

### CODE REVIEW

File: src/applications/upgradeable_contract/upgradeable_contract.md [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# Upgradeable Contract
+
+Starknet distinguishes between a contract and its implementation by separating contracts into classes and instances.
+
+A contract class is the definition of the contract: Cairo byte code, hint information, entry point names, and everything that defines its semantics unambiguously.
+
+Each class is identified by its class hash, which is analogous to a class name in an object-oriented programming language. A contract instance is a deployed contract corresponding to a class.
```
Comment (julio4) [inline src/applications/upgradeable_contract/upgradeable_contract.md:7]: As I suggested to add the POO comparison right at the start, maybe remove/change it there:
"To identify different contract classes, Starknet assigns a unique identifier to each class: the **class hash**. A contract instance is a deployed contract that corresponds to a specific contract class. Think of it as an instance of an object in languages like Java or Python."

File: src/applications/upgradeable_contract/upgradeable_contract.md [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+You can upgrade the class hash of a deployed contract by calling the `replace_class_syscall`. This example shows two contracts and how you can make them upgradeable. Try to deploy `UpgradeableContract_V0`, and then upgrade it to `UpgradeableContract_V1`.
```
Comment (julio4) [inline src/applications/upgradeable_contract/upgradeable_contract.md:9]: I would add another sentence after the first one to emphasize on the fact that this is what allow to upgrade the implementation of a contract:
"You can upgrade a deployed contract to a newer version by calling the `replace_class_syscall` function. By using this function, you can update the class hash associated with a deployed contract, effectively upgrading its implementation."

Comment (julio4) [reply src/applications/upgradeable_contract/upgradeable_contract.md:9]: And for the last part detailing the example, maybe start as a new paragraph to separate the theory and the example explanation. I would go further in the explanation to ensure better comprehension, like this:
"To illustrate this concept, let's consider an example with two contracts: `UpgradeableContract_V0` and `UpgradeableContract_V1`. Begin by deploying `UpgradeableContract_V0` as the initial version. Next, call `replace_class_syscall` to upgrade the class hash of the deployed contract to the one refering to the `UpgradeableContract_V1` implementation. Then, call the `version` function to see that the contract was upgraded to the V1 version."

**Review by julio4 — COMMENTED** [review]
This is a clear and easy-to-understand explanation. I added some suggestions for providing more details to (try to) further clarify the topic. Let me know if you like the style/tone.

**Comment by enitrat** [comment]
Thanks for the valuable feedback @julio4 !

**Pull request merged**